File: src/main/java/no/novari/flyt/mapping/model/configuration/FromCollectionMapping.java
package no.novari.flyt.mapping.model.configuration;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

import javax.validation.constraints.NotBlank;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class FromCollectionMapping<T> {

    public FromCollectionMapping(List<String> instanceCollectionReferencesOrdered, T elementMapping) {
        this.instanceCollectionReferencesOrdered = Optional.ofNullable(instanceCollectionReferencesOrdered).orElse(new ArrayList<>());
        this.elementMapping = elementMapping;
    }

    private final List<@NotBlank String> instanceCollectionReferencesOrdered;

    private final T elementMapping;

}

File: src/main/kotlin/no/novari/flyt/mapping/model/configuration/FromCollectionMapping.kt
package no.novari.flyt.mapping.model.configuration

import jakarta.validation.constraints.NotBlank

data class FromCollectionMapping<T>(
    val instanceCollectionReferencesOrdered: List<@NotBlank String> = emptyList(),
    val elementMapping: T,
) {
    class Builder<T> {
        private var instanceCollectionReferencesOrdered: List<String>? = null
        private var elementMapping: T? = null

        fun instanceCollectionReferencesOrdered(instanceCollectionReferencesOrdered: List<String>) =
            apply { this.instanceCollectionReferencesOrdered = instanceCollectionReferencesOrdered }

        fun elementMapping(elementMapping: T) = apply { this.elementMapping = elementMapping }

        fun build(): FromCollectionMapping<T> {
            return FromCollectionMapping(
                instanceCollectionReferencesOrdered = instanceCollectionReferencesOrdered ?: emptyList(),
                elementMapping = requireNotNull(elementMapping) { "elementMapping is required" },
            )
        }
    }

    companion object {
        @JvmStatic
        fun <T> builder(): Builder<T> = Builder()
    }
}
